### PR TITLE
BTreeMap: lightly refactor range_search

### PR DESCRIPTION
**Pull request by ssomers**
Ramp up to #81094 but IMHO useful in itself.

r? @Mark-Simulacrum

### CODE REVIEW

File: library/alloc/src/collections/btree/navigate.rs [hunk 1]
```diff
@@ -42,50 +42,46 @@ where
 
     let mut min_node = root1;
     let mut max_node = root2;
-    let mut min_found = false;
-    let mut max_found = false;
+    let mut min_bound = Some(start);
+    let mut max_bound = Some(end);
```
Comment (Mark-Simulacrum) [inline library/alloc/src/collections/btree/navigate.rs:46]: Can we place a comment on what the meaning of Some/None here is?

Comment (ssomers) [reply library/alloc/src/collections/btree/navigate.rs]: How about a commented type?

@rustbot modify labels: -S-waiting-on-author +S-waiting-on-review

**Comment by Mark-Simulacrum** [comment]
Looking a bit more at this, I am struggling with seeing this as an improvement. It seems like the encoding of the bit true/false into the Option makes it less clear what the code is doing rather than more by combining the paths for initial unbounded with "became unbounded" in a way that's confusing to me at least.

Maybe we can find a different structure that works well? Maybe noting on the various branches why first/last/left/right edges are chosen would also help with making the intent clear, right now I at least have a bit of a hard time justifying without some manual working out why they're being made.

**Comment by ssomers** [comment]
I can only repeat I think it's slightly better in itself and that #81094 is the next step, which splits this large function up, and is also necessary for the `drain` proposal. If you have something concrete to alter, I can do that. If you want to skip this step, fine by me. If you want to skip the rest of `drain`, that's also fine.

**Comment by ssomers** [comment]
The only thing I can think of that doesn/t thwart #81094 is the definition of `NoneableBound`: name, whether it is a proper enum,  reusing Bound or redefiningm perhaps more clearly distinguishing "include none" from unbounded.

**Comment by ssomers** [comment]
Can't keep updating this preparation of a preparation

**Pull request closed without merging**